### PR TITLE
disttask: handle left subtask in running state

### LINKED ISSUE #46735: import into jobs is always running after kill ddl owner
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
1、tidb_enable_dist_task='on'
2、import into user_data1 from 's3://brie/mvindex-10000000-csv/test.user_data1.*.csv?access-key=minioadmin&secret-access-key=minioadmin&endpoint=http://minio.pingcap.net:9001&force-path-style=true'  WITH skip_rows=1
3、after 2mins，kill ddl owner

### 2. What did you expect to see? (Required)
import into task can finish

### 3. What did you see instead (Required)
import into jobs is always running
mysql> select * from mysql.tidb_import_jobs where start_time >= '2023-09-06 17:20:09';
+--------+----------------------------+----------------------------+----------------------------+----------+--------------+------------+----------+------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------+---------+-----------+---------+---------------+
| id     | create_time                | start_time                 | update_time                | end_time | table_schema | table_name | table_id | created_by | parameters                                                                                                                                                                                                                                         | source_file_size | status  | step      | summary | error_message |
+--------+----------------------------+----------------------------+----------------------------+----------+--------------+------------+----------+------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------+---------+-----------+---------+---------------+
| 270002 | 2023-09-06 17:20:10.690458 | 2023-09-06 17:20:11.245575 | 2023-09-06 17:20:11.245575 | NULL     | importtest   | user_data1 |      539 | root@%     | {"file-location":"s3://brie/mvindex-10000000-csv/test.user_data1.*.csv?access-key=xxxxxx\u0026endpoint=http%3A%2F%2Fminio.pingcap.net%3A9001\u0026force-path-style=true\u0026secret-access-key=xxxxxx","format":"csv","options":{"skip_rows":"1"}} |       1995557384 | running | importing | NULL    | NULL          |
+--------+----------------------------+----------------------------+----------------------------+----------+--------------+------------+----------+------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------+---------+-----------+---------+---------------+
1 row in set (0.01 sec)


### 4. What is your TiDB version? (Required)
./tidb-server -V
 Release Version: v7.4.0-alpha
Edition: Community
Git Commit Hash: 66fe053b468e8d4c1f182b98e796ca48075809a7
Git Branch: heads/refs/tags/v7.4.0-alpha
UTC Build Time: 2023-09-05 11:43:48
GoVersion: go1.21.0
Race Enabled: false
Check Table Before Drop: false
Store: unistore

### PR DESCRIPTION
**Pull request by D3Hunter**
### What problem does this PR solve?

Issue Number: close #46735

Problem Summary:

### What is changed and how it works?

handle left subtask in running state depends on whether the subtask is idempotent, if true we rerun it, else we fail it.

all import-subtask is idempotent. add-index is left as true too, but require additional pr to make real true.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: ddl/backfilling_dist_scheduler.go [hunk 1]
```diff
@@ -160,6 +160,11 @@ func (s *backfillDistScheduler) GetSubtaskExecutor(ctx context.Context, task *pr
 	}
 }
 
+func (*backfillDistScheduler) IsIdempotent(*proto.Subtask) bool {
+	// TODO: implement this later.
+	return false
```
Comment (D3Hunter) [inline ddl/backfilling_dist_scheduler.go:165]: @Benjamin2037 @tangenta ptal

File: ddl/backfilling_dist_scheduler.go [hunk 2]
```diff
@@ -160,6 +163,12 @@ func (s *backfillDistScheduler) GetSubtaskExecutor(ctx context.Context, task *pr
 	}
 }
 
+func (s *backfillDistScheduler) IsIdempotent(*proto.Subtask) bool {
+	// unique index depends on MVCC to do deduplicate, so it's not idempotent.
+	// we might save engine TS later to make it idempotent.
+	return !s.uniqueIndex
```
Comment (Benjamin2037) [inline ddl/backfilling_dist_scheduler.go:169]: Currently, the unique index should also be idempotent.

File: disttask/framework/scheduler/scheduler.go [hunk 3]
```diff
@@ -198,10 +200,23 @@ func (s *BaseScheduler) run(ctx context.Context, task *proto.Task) error {
...
 
-		s.startSubtaskAndUpdateState(subtask)
-		if err := s.getError(); err != nil {
-			logutil.Logger(s.logCtx).Warn("startSubtaskAndUpdateState meets error", zap.Error(err))
-			continue
+		if subtask.State == proto.TaskStateRunning {
+			if !s.IsIdempotent(subtask) {
+				logutil.Logger(s.logCtx).Info("subtask in running state and is not idempotent, fail it",
+					zap.Int64("subtask-id", subtask.ID))
+				subtaskErr := errors.New("subtask in running state and is not idempotent")
```
Comment (tangenta) [inline disttask/framework/scheduler/scheduler.go:207]: Maybe assign an error code and add it to `ReorgRetryableErrCodes`?

File: ddl/backfilling_dist_scheduler.go (hunk 2 above)
Comment (D3Hunter) [reply ddl/backfilling_dist_scheduler.go:169]: as we have discussed, will mark all add-index subtask as Idempotent

**Review by tangenta — APPROVED** [review]

**Review by Benjamin2037 — APPROVED** [review]
LGTM

**Pull request merged**